app/controllers/order_items_controller.rb

```diff
@@ -0,0 +1,82 @@
+class OrderItemsController < ApplicationController
+  before_action :check_if_product_is_in_cart, only: [:create]
+
+  def create
+    @order = current_order
+    @order_item = @order.order_items.create(order_item_params)
+    if @order_item.save
+      redirect_to cart_path
+    else
+      if @order_item.product.inventory == 1
+        flash[:error] = "Sorry, there is only #{@order_item.product.inventory} #{@order_item.product.name} available."
+      elsif @order_item.product.inventory == 0
+        flash[:error] = "Sorry, there are no #{@order_item.product.name.pluralize} available."
+      else
+        flash[:error] = "Sorry, there are only #{@order_item.product.inventory} #{@order_item.product.name.pluralize} available."
+      end
+      redirect_to product_path(@order_item.product)
+    end
+  end
+
+  def destroy
+    @order = current_order
+    @order_item = @order.order_items.find(params[:id])
+    @order_item.destroy
+    @order_items = @order.order_items
+
+    redirect_to cart_path
+  end
+
+  def update
+    @order_item = OrderItem.find(params[:id])
+    @order_item.update_attributes(order_item_params)
+    if @order_item.save
+      redirect_to cart_path
+    else
+      flash[:error] = "Sorry, there are only #{@order_item.product.inventory} #{@order_item.product.name.pluralize} available."
+      redirect_to cart_path
+    end
+  end
+
+  def check_if_product_is_in_cart
+    order = current_order
+    @product_id = params[:order_item][:product_id]
+    product_in_cart = order.order_items.select {|order_item| order_item.product_id == @product_id.to_i}
+
+    if !product_in_cart.empty?
+      update_product_in_cart(product_in_cart[0])
+    end
+  end
+
+  def update_product_in_cart(product_in_cart)
+    current_quantity = product_in_cart.quantity
+    additional_quantity = params[:order_item][:quantity].to_i
+    params[:order_item][:quantity] = current_quantity + additional_quantity
+
+    product_in_cart.update_attributes(order_item_params)
```

Comment:
I'm a bit concerned that this call to `update_attributes` might be _too_ flexible. This code, I think, will run when a guest attempts to add an item that is already within their cart.

In that case we should not allow them to mark their item as shipped, for example. However, by using `order_item_params` with `update_attributes` we're allowing the web browser to send any of these parameters: `order_id, product_id, quantity, status`.

Even though our views don't have a form which asks guests for the OrderItem's status on their cart items, a knowledgeable user could provide the `status` parameter directly, and it looks like our code would happily accept it.

Instead, I'd advise creating a second method like `order_item_params` that was specific to the guests (like `order_item_guest_params`), or specific to a particular controller action (like `order_item_create_params`). This version would only include the params we need to identify the order item (`order_id` and `product_id`) plus the params we should allow the user to set (`quantity`).